osu.Game/Skinning/Components/BeatmapAttributeText.cs

```diff
@@ -123,6 +145,132 @@ private void updateLabel()
+                return;
+
+            double rate = ModUtils.CalculateRateWithMods(mods.Value);
+
+            valueDictionary[BeatmapAttribute.Length] = TimeSpan.FromMilliseconds(beatmapInfo.Length / rate).ToFormattedDuration();
+            valueDictionary[BeatmapAttribute.BPM] = FormatUtils.RoundBPM(beatmapInfo.BPM, rate).ToLocalisableString(@"F0");
+        }
+
+        private static bool isMetadata(BeatmapAttribute attribute) => (attribute >= BeatmapAttribute.Title && attribute <= BeatmapAttribute.Creator)
+                                                                      || attribute == BeatmapAttribute.RankedStatus || attribute == BeatmapAttribute.Source;
+
+        private static bool isBeatmapDifficultyAttribute(BeatmapAttribute attribute) => (attribute <= BeatmapAttribute.ApproachRate);
+
+        private static bool isTimingData(BeatmapAttribute attribute) => attribute == BeatmapAttribute.Length || attribute == BeatmapAttribute.BPM;
+
+        private static bool isDiffcalcInfo(BeatmapAttribute attribute) => attribute == BeatmapAttribute.StarRating || attribute == BeatmapAttribute.MaxPerformance;
+
+        #region diffcalc stuff
+
+        [Resolved]
+        private BeatmapDifficultyCache difficultyCache { get; set; } = null!;
+
+        private Bindable<StarDifficulty?>? bindableDifficulty;
+
+        private CancellationTokenSource? cancellationTokenSource;
+
+        private void updateBindableDifficulty()
+        {
+            cancellationTokenSource?.Cancel();
+            bindableDifficulty?.UnbindEvents();
+
+            // Skip if not relevant attribute
+            if (!isDiffcalcInfo(Attribute.Value))
+                return;
+
+            cancellationTokenSource = new CancellationTokenSource();
+
+            bindableDifficulty = (Bindable<StarDifficulty?>)difficultyCache.GetBindableDifficulty(beatmapInfo, cancellationTokenSource.Token);
+            bindableDifficulty.BindValueChanged(d =>
+            {
+                StarDifficulty difficulty = d.NewValue ?? new StarDifficulty();
+
+                if (Attribute.Value == BeatmapAttribute.StarRating)
+                {
+                    valueDictionary[BeatmapAttribute.StarRating] = difficulty.Stars.ToLocalisableString(@"F2");
+                    updateLabel();
+                }
+                else if (Attribute.Value == BeatmapAttribute.MaxPerformance)
+                {
+                    calculateMaxPerformance(difficulty.Attributes).ContinueWith(t =>
+                    {
+                        valueDictionary[BeatmapAttribute.MaxPerformance] = t.GetResultSafely().ToLocalisableString(@"0pp");
+                        updateLabel();
+                    }, cancellationTokenSource.Token);
+                }
+            });
+        }
+
+        private async Task<double> calculateMaxPerformance(DifficultyAttributes? difficultyAttributes)
+        {
+            if (difficultyAttributes == null || cancellationTokenSource == null)
```

Comment:
huh? what's is this supposed to imply?

Reply:
sanity check so it will not try to calculate pp if something happened during `difficultyAttributes` calculation (either cancelled or error)